apps/antalmanac/src/lib/enrollmentHistory.ts

```diff
@@ -0,0 +1,154 @@
+import { queryGraphQL } from './helpers';
+import { termData } from './termData';
+
+// This represents the enrollment history of a course section during one quarter
+export interface EnrollmentHistoryGraphQL {
+    year: string;
+    quarter: string;
+    department: string;
+    courseNumber: string;
+    dates: string[];
+    totalEnrolledHistory: string[];
+    maxCapacityHistory: string[];
+    waitlistHistory: string[];
+    instructors: string[];
+}
+
+export interface EnrollmentHistoryGraphQLResponse {
+    data: {
+        enrollmentHistory: EnrollmentHistoryGraphQL[];
+    };
+}
+
+// To organize the data and make it easier to graph the enrollment
+// data, we can merge the dates, totalEnrolledHistory, maxCapacityHistory,
+// and waitlistHistory arrays into one array that contains the enrollment data
+// for each day
+export interface EnrollmentHistory {
+    year: string;
+    quarter: string;
+    department: string;
+    courseNumber: string;
+    days: EnrollmentHistoryDay[];
+    instructors: string[];
+}
+
+export interface EnrollmentHistoryDay {
+    date: string;
+    totalEnrolled: number;
+    maxCapacity: number;
+    waitlist: number | null;
+}
+
+class _EnrollmentHistory {
+    // Each key in the cache will be the department and courseNumber concatenated
+    enrollmentHistoryCache: Record<string, EnrollmentHistory>;
+    termShortNames: string[];
+
+    constructor() {
+        this.enrollmentHistoryCache = {};
+        this.termShortNames = termData.map((term) => term.shortName);
+    }
+
+    queryEnrollmentHistory = async (department: string, courseNumber: string): Promise<EnrollmentHistory | null> => {
+        const cacheKey = department + courseNumber;
+        if (cacheKey in this.enrollmentHistoryCache) {
+            return this.enrollmentHistoryCache[cacheKey];
+        }
+
+        // Query for the enrollment history of all lecture sections that were offered
+        const queryString = `{
+            enrollmentHistory(department: "${department}", courseNumber: "${courseNumber}", sectionType: Lec) {
+                year
+                quarter
+                department
+                courseNumber
+                dates
+                totalEnrolledHistory
+                maxCapacityHistory
+                waitlistHistory
+                instructors
+            }
+        }`;
```

Comment:
I wonder how efficient it would be store the template string in a more global scope, and then apply immutable string manipulation. This is extremely minute; it's just worth considering what scope is appropriate for variables that don't need to be initialized multiple times.

```tsx
const QUERY_TEMPLATE = `{
        enrollmentHistory(department: "$$DEPARTMENT$$", courseNumber: "$$COURSE$$", sectionType: Lec) {
            year
            quarter
            department
            courseNumber
            dates
            totalEnrolledHistory
            maxCapacityHistory
            waitlistHistory
            instructors
        }
    }`;

// ...
const queryString = QUERY_TEMPLATE.replace('$$DEPARTMENT', ...).replace('$$COURSE$$', ...)
```

Reply:
Still not sure about the efficiency, but I do think putting it in a more visible scope is cleaner. For now I just have it as a static variable in `DepartmentEnrollmentHistory`, but I also see the case to put it globally.